### ISSUE TITLE
[bug] Too many delete_obsolete_objects tasks

### ISSUE THREAD

**Issue by nemesisdesign** [openwisp/openwisp-notifications#163]
In production systems, I observe too many `delete_obsolete_objects` being executed, which is unnecessary.

We should find a way to limit these.

One possibility would be to allow registering models which are sensitive to the pre_delete signals and allow each module to add their models, so we'd have:

- User model added by default by openwisp-notifications
- Device model added by default by openwisp-controller

These at the moment are the only two modules we need to listen to pre-delete events for.

**Comment by pandafy** [comment]
I don't like burdening developers with the task of registering models to perform housekeeping operations. We should find a dynamic method of registering these signals. This is what I have in mind: 

**For every referenced object in a notification, connect the `related_object_deleted` function as receiver to the `pre_delete` signal of the model of the referenced object.**

**Implementation**

1. Maintain a list of names of models(preferably `app_label.model_name`) for which receiver for `pre_delete` signal has been registered. Let's reference this list by `pre_delete_sensitive_models`
2. While creating a notification, check whether the model names of all three referenced objects(foreign keys), exists in above list. 
3. If it does not, connect a new receiver to the `pre_delete` signal as following

```python
from django.db.models.signals import pre_delete

pre_delete.connect(receiver=related_object_deleted, sender=model, weak=False, dispatch_uid=f'{model._meta.app_label}_{model._meta.model_name}_pre_delete' )
```
4. Update `pre_delete_sensitive_models` with this model. 


**CAVEAT:** Additional computational while creating a new notification to traverse `pre_delete_sensitive_models`. But length of this list would be always small.

**Comment by nemesisdesign** [comment]
> * Maintain a list of names of models(preferably `app_label.model_name`) for which receiver for `pre_delete` signal has been registered. Let's reference this list by `pre_delete_sensitive_models`

How do you want to generate this list of models? Do we already have this info?

I looked at the register functions but it doesn't look like we do.
We could add it there though, we should just make sure that the result is a set, not a list, so we avoid duplicates (since we have multiple registrations on the same models).

> * While creating a notification, check whether the model names of all three referenced objects(foreign keys), exists in above list.
> * If it does not, connect a new receiver to the `pre_delete` signal as following

Ok! The uuid should probably mention `ns_cache_invalidation` to avoid collisions, since `pre_delete` is very generic and could be used elsewhere.

There's a caveat: multiple signals could be used, as in #164, we need a dedicated signal for that and we will need to hook other dedicated signals in the future for other objects.

We can't use the `change` signal for models like Device or others because their information is updated often. There's probably a bug there too (for some reason the Device model is sending `change` signals more often than what is necessary), but I still haven't been able to replicate it. However, notwithstanding this probable bug in Device model, if we want to make this mechanism efficient, we should provide a way to configure this behavior, since every model in the application will be different and I expect we'll be adding many new notifications within the next 2 years, eg: network topology and radius. 
Some objects do not get updated often and there is fine to hook to the change signal, but other models do change often (network topology links, nodes and radius accounting sessions) so in those cases dedicated signals are necessary (network topology already has some dedicated signals because of this).

In short: I agree in make sure it's maintainable for us, but I am more concerned with making it efficient, because as the number of users and devices in the system grows, if these things are not tuned to max efficiently, a small inefficiency can cause the background worker to halt, and it's something you won't see during development but only in prod, so the system must be designed in a way that it can be fined tuned.
